server/backup/backup_s3.go

```diff
@@ -61,27 +51,123 @@ func (s *S3Backup) Remove() error {
 	return os.Remove(s.Path())
 }
 
+// Reader provides a wrapper around an existing io.Reader
+// but implements io.Closer in order to satisfy an io.ReadCloser.
+type Reader struct {
+	io.Reader
+}
+
+func (Reader) Close() error {
+	return nil
+}
+
 // Generates the remote S3 request and begins the upload.
-func (s *S3Backup) generateRemoteRequest(rc io.ReadCloser) (*http.Response, error) {
-	r, err := http.NewRequest(http.MethodPut, s.PresignedUrl, nil)
+func (s *S3Backup) generateRemoteRequest(rc io.ReadCloser) error {
+	defer rc.Close()
+
+	size, err := s.Backup.Size()
 	if err != nil {
-		return nil, err
+		return err
 	}
 
-	if sz, err := s.Size(); err != nil {
-		return nil, err
-	} else {
-		r.ContentLength = sz
-		r.Header.Add("Content-Length", strconv.Itoa(int(sz)))
+	urls, err := api.New().GetBackupRemoteUploadURLs(s.Backup.Uuid, size)
+	if err != nil {
+		return err
+	}
+
+	log.Debug("attempting to upload backup to remote S3 endpoint")
+
+	handlePart := func(part string, size int64) (string, error) {
+		r, err := http.NewRequest(http.MethodPut, part, nil)
+		if err != nil {
+			return "", err
+		}
+
+		r.ContentLength = size
+		r.Header.Add("Content-Length", strconv.Itoa(int(size)))
 		r.Header.Add("Content-Type", "application/x-gzip")
+
+		r.Body = Reader{io.LimitReader(rc, size)}
+
+		res, err := http.DefaultClient.Do(r)
+		if err != nil {
+			return "", err
+		}
+
+		defer res.Body.Close()
+
+		if res.StatusCode != http.StatusOK {
+			return "", fmt.Errorf("failed to put S3 object part, %d:%s", res.StatusCode, res.Status)
+		}
+
+		return res.Header.Get("ETag"), nil
 	}
 
-	r.Body = rc
+	// Keep track of errors from individual part uploads.
+	hasError := true
+	defer func() {
+		if !hasError {
+			return
+		}
+
+		r, err := http.NewRequest(http.MethodPost, urls.AbortMultipartUpload, nil)
+		if err != nil {
+			log.WithError(err).Warn("failed to create http request (AbortMultipartUpload)")
+			return
+		}
+
+		res, err := http.DefaultClient.Do(r)
```

Comment:
naughty naughty, this can hang indefinitely due to the default client not having a timeout associated with it.